### ISSUE TITLE
torch.all(input, dim, out) ：When out.dtype == torch.uint8, there is a calculation error.

### ISSUE THREAD

**Issue by yangyang-01-debug** [pytorch/pytorch#117215]
### 🐛 Describe the bug

```
import torch
info=torch.iinfo(torch.uint8)
a = torch.randint(info.min,info.max,(73,11,3,17), dtype=torch.uint8)
b = torch.all(a, dim=0)

c = a.to(torch.bool).all(dim=0)
print(torch.ne(b, c).sum())
# ‘b' and 'c' must be the same
```

```
a = torch.randn((73,11,3,17))
info=torch.iinfo(torch.uint8)
out = torch.randint(info.min,info.max,(11,3,17), dtype=torch.uint8)

torch.all(a, dim=0, out=out)
c = (a.to(torch.bool).all(dim=0))

print(torch.ne(out, c).sum())
# ‘b' and 'c' must be the same
```

### Versions

Collecting environment information...
PyTorch version: 2.0.0a0+gitd02fa3c
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 7.5.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.17

Python version: 3.8.18 (default, Sep 11 2023, 13:40:15)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                256
On-line CPU(s) list:   0-255
Thread(s) per core:    2
Core(s) per socket:    64
座：                 2
NUMA 节点：         2
厂商 ID：           AuthenticAMD
CPU 系列：          23
型号：              49
型号名称：        AMD EPYC 7H12 64-Core Processor
步进：              0
CPU MHz：             1500.000
CPU max MHz:           2600.0000
CPU min MHz:           1500.0000
BogoMIPS：            5199.85
虚拟化：           AMD-V
L1d 缓存：          32K
L1i 缓存：          32K
L2 缓存：           512K
L3 缓存：           16384K
NUMA 节点0 CPU：    0-63,128-191
NUMA 节点1 CPU：    64-127,192-255
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc art rep_good nopl nonstop_tsc extd_apicid aperfmperf eagerfpu pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_l2 cpb cat_l3 cdp_l3 hw_pstate sme retpoline_amd ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip overflow_recov succor smca

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] torch==2.0.0a0+gitd02fa3c
[pip3] triton==2.1.0
[conda] blas                      1.0                         mkl    http://jfrog.tecorigin.net/artifactory/api/conda/tecotp-vir_conda
[conda] mkl                       2023.1.0         h213fc3f_46343    http://jfrog.tecorigin.net/artifactory/api/conda/tecotp-vir_conda
[conda] mkl-include               2023.1.0         h06a4308_46343    http://jfrog.tecorigin.net/artifactory/api/conda/tecotp-vir_conda
[conda] mkl-service               2.4.0            py38h5eee18b_1    http://jfrog.tecorigin.net/artifactory/api/conda/tecotp-vir_conda
[conda] mkl_fft                   1.3.8            py38h5eee18b_0    http://jfrog.tecorigin.net/artifactory/api/conda/tecotp-vir_conda
[conda] mkl_random                1.2.4            py38hdb19cb5_0    http://jfrog.tecorigin.net/artifactory/api/conda/tecotp-vir_conda
[conda] numpy                     1.23.0                   pypi_0    pypi
[conda] torch                     2.0.0a0+gitd02fa3c           dev_0    <develop>
[conda] triton                    2.1.0                    pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kadeng

**Comment by colesbury** [comment]
I can reproduce the issue on colab

**Comment by ezyang** [comment]
This is probably the thing where sometimes the bool kernels expect a canonical form for the uint8 representing the bool (it's either 0 or 1) but sometimes we violate this invariant and so weird things happen

**Comment by colesbury** [comment]
This looks like a simple typo. It should be a logical `&&` not bitwise `&`. The other parts of the kernel correctly use `&&`.

https://github.com/pytorch/pytorch/blob/6d1406d177c5b4c0e627830c12b94619380aa6aa/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L294

**Comment by colesbury** [comment]
There's a similar bug in `or_kernel`

**Comment by ezyang** [comment]
@colesbury you gonna send us the fix?

**Comment by colesbury** [comment]
@ezyang, no I was not planning on it.

**Comment by ezyang** [comment]
A little trickier because Vectorized doesn't support &&. Not too bad though.

**Comment by vineethsaivs** [comment]
The second repro in this report still reproduces on 2.11.0. Only the first one was fixed, so I think this was closed a little early.

```
### repro 1 (uint8 input, no out=)
   mismatches: 0            -> fixed

### repro 2 (float input, uint8 out=)
   mismatches: 561 of 561   -> still broken
```

Reduced, it is not about randomness or size at all:

```python
>>> import torch
>>> x = torch.tensor([256, 512])
>>> torch.all(x).item()
True
>>> torch.all(x, out=torch.empty((), dtype=torch.uint8)).item()
0
>>> torch.all(x, out=torch.empty((), dtype=torch.bool)).item()
True
```

So passing `out=` alone flips the answer, and `torch.any` behaves the same way. A `uint8` out is explicitly permitted here (`check_result_is_bytebool` accepts Bool and Byte and rejects Long), so this is a supported call rather than misuse.

The cause looks like the CPU branch of `get_allany_iter` in `aten/src/ATen/native/ReduceOps.cpp`, which passes `result.scalar_type()` as the reduction's **input** dtype:

```cpp
meta::make_reduction_from_out_ty(self, result, dims, keepdim, result.scalar_type());
```

`make_reduction` then does `self.to(in_dtype)`, so a `uint8` out casts the *input* to uint8 before `and_stub`/`or_stub` runs, and `torch.tensor([256, 512]).to(torch.uint8)` is `[0, 0]`. Any value that truncates or wraps to zero (`256`, `512`, `0.5`) is silently read as False. The comment above that line says the overload "casts the input to kBool", which is only true when the out tensor is bool. The CUDA/MPS branch one line up passes `self.scalar_type()` and is unaffected, so CPU and CUDA disagree.

Two things that suggest it is unintended rather than a quirk of the uint8 path. The `numel() == 1` fast path in `allany_impl` does `result.copy_(self.view_as(result).to(at::kBool))` and is correct, which gives an internal contradiction:

```python
>>> torch.all(torch.tensor([256]), out=torch.empty((), dtype=torch.uint8)).item()       # numel==1 fast path
1
>>> torch.all(torch.tensor([256, 512]), out=torch.empty((), dtype=torch.uint8)).item()  # general path
0
```

And *Note [all, any : uint8 compatibility]* says the uint8 path exists "For NumPy compatibility", while NumPy gives the opposite answer:

```python
>>> np.all(np.array([256, 512]), out=np.empty((), dtype=np.uint8)).item()
1
```

It also affects the `dim=` overload with float input, where only some entries are wrong, which is the harder shape to notice:

```python
>>> x = torch.tensor([[0.5, 0.5], [2.0, 3.0]])
>>> torch.all(x, 1).tolist()
[True, True]
>>> torch.all(x, 1, out=torch.empty(0, dtype=torch.uint8)).tolist()
[0, 1]
```

Worth noting the existing regression test does not cover this: `test_all_any_vs_numpy._test_out_variant` only ever uses `out = torch.empty_like(x)`, so the out dtype always equals the input dtype and the mixed combination is never exercised.

Could this be reopened, and marked `actionable` if you agree it should be fixed? I have a fix in mind (canonicalise the reduction input to a byte-width truth value instead of to the out dtype, keeping the uint8 result, or reduce into a bool temporary and copy as the `numel()==1` path already does) plus the test extension above, and I would rather have the issue open before sending a PR, per the contribution guide. I do not have a build environment for the C++ change, so I would want CI to be the runner and would say so on the PR.
